### DISCUSSION TITLE
not able to use com.apple.developer.storekit.external-purchase entitlement

### DISCUSSION THREAD

**Discussion by jesseblom** [expo/expo#30976] [Archived: Questions]
Hey,
We have an external pay provider for our iOS app and ive added the following part to our app.json:
```
    "ios": {
      "supportsTablet": true,
      "buildNumber": "1",
      "bundleIdentifier": "com.redacted",
      "entitlements": {
        "com.apple.developer.storekit.external-purchase-link": "true"
```

But xcode logging from expo gives me an error about it and tells me the provisioning profile doesn't include the in-app purchase, while in fact it does. 

My question is if the entitlement from apple, which is fairly recent, is already implemented and able to be used?

It is not listed in https://docs.expo.dev/build-reference/ios-capabilities/ 
and on top of that it's also not even listed in https://developer.apple.com/help/account/reference/supported-capabilities-ios

But to submit our app we need to use this storekit since the recent european changes (july/august) in order to use an external payment provider.

**Comment by CoreyBovalina** [comment]
It works if Apple has given your app approval to use it.

**Comment by loorko** [comment]
@jesseblom @CoreyBovalina 
It's only loosely related to the topic, but you could tell me
how did you implement it in React Native the `await ExternalPurchase.canPresent` and `await ExternalPurchase.presentNoticeSheet()`  function call?

**Reply by CoreyBovalina** [reply]
I created a local native module and in the swift code.  I check to make sure it can open, if it can't I present a modal based off Apple's specifications that are similar to if it returned true, otherwise Apple offers the modal up natively. Looks something like this:
```
import ExpoModulesCore
import StoreKit

public class ExternalLinkModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExternalLink')` in JavaScript.
    Name("ExternalLink")

    AsyncFunction("canOpenExternalLinkHelper") {(promise:Promise) in
      Task{
          // this will return false in case the website link is not available in Info.plist
          guard SKPaymentQueue.canMakePayments() else {
              return promise.resolve(false)
          }
          if #available(iOS 16.0, *) {
            if(await ExternalLinkAccount.canOpen){
              return promise.resolve(true)
            }
          }
          return promise.resolve(false)
      }
    }


    AsyncFunction("openExternalLinkHelper") { (promise:Promise) in
      Task{
              if #available(iOS 16.0, *) {
                  do {
                      try await ExternalLinkAccount.open()
                      return promise.resolve(true)
                  } catch {
                      dump("Error - \(error)")
                  }
              } else {
                  return promise.resolve(false)
              }
      }
    }
  }
}
```


```
// Import the native module. 
// and on native platforms to ExternalLink.ts
import ExternalLinkModule from './src/ExternalLinkModule';

export const openExternalLinkHelper = async () => {
  return await ExternalLinkModule.openExternalLinkHelper();
};

export const canOpenExternalLinkHelper = async () => {
  return await ExternalLinkModule.canOpenExternalLinkHelper();
};
```

**Reply by loorko** [reply]
@CoreyBovalina 
Based on your suggestion, I created my own solution, which is also available as an npm package, but when I run it as developmentClient, I get an error:
**Unable to Complete Request**

- app.json is configured
`{
  "expo": {
    ...
    "ios": {
      "supportsTablet": true,
      "bundleIdentifier": "com.your.bundleIdentifier",
      "entitlements": {
        "com.apple.developer.storekit.external-purchase": true
      },
      "infoPlist": {
        "SKExternalPurchase": ["hu"]
      }
    }
    ...
  }
}`
- In the **App ID Configuration** / **Additional Capabilities** tab, the **StoreKit External Purchase** is checked
- Provisioning Profile file updated

What could be the problem?
How could I debug this?
Is it possible that it would work in a production environment?

Source code:
https://github.com/loorko/expo-external-purchase/blob/master/ios/ExpoExternalPurchaseModule.swift

NPM package:
https://www.npmjs.com/package/expo-external-purchase

**Reply by CoreyBovalina** [reply]
@loorko the entitlement has to be available for that bundle identifier. 

in my case I have a function that checks the build profile and sets that entitlement:

` const getExternalLink = () => {
    if (buildProfile === 'production') {
      return { 'com.apple.developer.storekit.external-link.account': true };
    }
    return {};
  };`

` entitlements: {
        ...getExternalLink(),
        'aps-environment': true
}`

**Reply by loorko** [reply]
@CoreyBovalina thank you for the quick reply

**Comment by loorko** [comment]
@CoreyBovalina I hope you are know the answer for this two questions:

1;
The field for entering the App Store Server Notifications URL does not appear unless we add an In-App Purchase item or a Subscriptions item. How can we provide the two App Store Server Notifications URLs for external purchases as described in the following documentation?
[https://developer.apple.com/help/app-store-connect/configure-in-app-purchase-settings/enter-server-urls-for-app-store-server-notifications
](https://developer.apple.com/help/app-store-connect/configure-in-app-purchase-settings/enter-server-urls-for-app-store-server-notifications
)
2;
Even though I set it to aps-environment: production, which is also visible in the app store connect Build Metadata Store Information section, but we are still receiving a SANDBOX identifie.
[https://developer.apple.com/documentation/bundleresources/entitlements/aps-environment/
](https://developer.apple.com/documentation/bundleresources/entitlements/aps-environment/
)
Thank you in advance for your help.
